File: Rerun.r
library(readxl)
library(writexl)
library(ggplot2)
library(dplyr)
library(tidyr)
library(stringr)
library(purrr)
library(scales)
library(gridExtra)
library(tools)

# Calculate amortization rate based on future capex and LOM ounces
calculate_amortization_rate <- function(future_capex, lom_ounces) {
  if (lom_ounces == 0) {
    return(NA)  # Avoid division by zero
  }
  return(as.numeric(future_capex) / as.numeric(lom_ounces))
}

# Calculate expected expense by multiplying amortization rate by ounces mined
calculate_expected_expense <- function(amortization_rate, ounces_mined) {
  return(as.numeric(amortization_rate) * as.numeric(ounces_mined))
}

# Perform sensitivity analysis on amortization rate and expected expenses
sensitivity_analysis <- function(base_future_capex, base_lom_ounces, 
                                ounces_mined = NULL, variation = 0.20, steps = 5) {
  # Calculate variation percentages
  percentages <- seq(-variation, variation, length.out = 2 * variation / (variation / steps) + 1)
  
  # Create vectors for future capex and LOM ounces variations
  future_capex_variations <- base_future_capex * (1 + percentages)
  lom_ounces_variations <- base_lom_ounces * (1 + percentages)
  
  # Create percentage labels for the table
  percentage_labels <- paste0(as.integer(percentages * 100), "%")
  
  # Initialize matrices for results
  amort_matrix <- matrix(0, nrow = length(percentage_labels), ncol = length(percentage_labels))
  
  # Fill the matrix with amortization rates
  for (i in 1:length(percentage_labels)) {
    for (j in 1:length(percentage_labels)) {
      amort_matrix[i, j] <- calculate_amortization_rate(future_capex_variations[i], 
                                                        lom_ounces_variations[j])
    }
  }
  
  # Create data frames from matrices
  amort_results <- as.data.frame(amort_matrix)
  colnames(amort_results) <- percentage_labels
  rownames(amort_results) <- percentage_labels
  
  # Calculate expected expenses if ounces mined is provided
  expense_results <- NULL
  if (!is.null(ounces_mined)) {
    expense_matrix <- matrix(0, nrow = length(percentage_labels), ncol = length(percentage_labels))
    for (i in 1:length(percentage_labels)) {
      for (j in 1:length(percentage_labels)) {
        expense_matrix[i, j] <- calculate_expected_expense(amort_matrix[i, j], ounces_mined)
      }
    }
    
    expense_results <- as.data.frame(expense_matrix)
    colnames(expense_results) <- percentage_labels
    rownames(expense_results) <- percentage_labels
  }
  
  return(list(amort_results = amort_results, expense_results = expense_results))
}

# Read input data from an Excel file
read_input_excel <- function(file_path) {
  tryCatch({
    # Try to read the Excel file
    df <- readxl::read_excel(file_path)
    
    # Validate that required columns exist
    required_columns <- c('Project', 'Future_Capex', 'LOM_Ounces', 'Ounces_Mined')
    missing_columns <- required_columns[!required_columns %in% colnames(df)]
    
    if (length(missing_columns) > 0) {
      cat("Error: Missing required columns in the Excel file:", paste(missing_columns, collapse = ", "), "\n")
      cat("Please ensure your Excel file has columns named: Project, Future_Capex, LOM_Ounces, Ounces_Mined\n")
      return(NULL)
    }
    
    # Ensure numeric columns are numeric
    df$Future_Capex <- as.numeric(df$Future_Capex)
    df$LOM_Ounces <- as.numeric(df$LOM_Ounces)
    df$Ounces_Mined <- as.numeric(df$Ounces_Mined)
    
    # Drop rows with NA values in required numeric columns
    df <- df[complete.cases(df[, c('Future_Capex', 'LOM_Ounces', 'Ounces_Mined')]), ]
    
    if (nrow(df) == 0) {
      cat("Error: No valid data rows found after cleaning.\n")
      return(NULL)
    }
    
    return(df)
  }, error = function(e) {
    cat("Error reading Excel file:", e$message, "\n")
    return(NULL)
  })
}

# Create and save a heatmap visualization
create_heatmap <- function(data, title, filename, save_dir) {
  # Reshape the data for ggplot
  plot_data <- as.data.frame(data) %>%
    rownames_to_column(var = "Future_Capex") %>%
    pivot_longer(cols = -Future_Capex, 
                names_to = "LOM_Ounces", 
                values_to = "Value")
  
  # Create the heatmap
  p <- ggplot(plot_data, aes(x = LOM_Ounces, y = Future_Capex, fill = Value)) +
    geom_tile() +
    geom_text(aes(label = sprintf("%.2f", Value)), color = "black", size = 3) +
    scale_fill_gradient(low = "lightyellow", high = "steelblue") +
    labs(title = title,
         x = "LOM Ounces Variation",
         y = "Future Capex Variation") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          plot.title = element_text(hjust = 0.5))
  
  # Save the plot
  filepath <- file.path(save_dir, filename)
  ggsave(filepath, p, width = 12, height = 10, dpi = 300)
  cat("Heatmap saved to:", filepath, "\n")
}

# Ask user for a file path and validate it exists
get_file_path <- function(prompt) {
  repeat {
    file_path <- trimws(readline(prompt))
    
    if (file_path == "") {
      return(NULL)
    }
    
    if (file.exists(file_path)) {
      return(file_path)
    } else {
      cat("Error: File '", file_path, "' does not exist.\n")
    }
  }
}

# Ask user for a directory path where files should be saved
get_save_path <- function() {
  repeat {
    save_dir <- trimws(readline("\nEnter the folder path to save results (or press Enter to use current directory): "))
    
    # Use current directory if input is empty
    if (save_dir == "") {
      return(getwd())
    }
    
    # Check if the directory exists
    if (dir.exists(save_dir)) {
      return(save_dir)
    } else {
      create_dir <- readline(paste0("Directory '", save_dir, "' doesn't exist. Create it? (y/n): "))
      if (tolower(create_dir) == 'y') {
        tryCatch({
          dir.create(save_dir, recursive = TRUE)
          cat("Created directory:", save_dir, "\n")
          return(save_dir)
        }, error = function(e) {
          cat("Error creating directory:", e$message, "\n")
        })
      } else {
        cat("Please enter a valid directory path.\n")
      }
    }
  }
}

# Main function
main <- function() {
  cat("Batch Sensitivity Analysis for Amortization Rate and Expected Expenses\n")
  cat("--------------------------------------------------------------------\n")
  
  # Get input Excel file path
  input_file <- get_file_path("Enter the path to your input Excel file: ")
  if (is.null(input_file)) {
    cat("Operation cancelled.\n")
    return(invisible(NULL))
  }
  
  # Read input data
  input_data <- read_input_excel(input_file)
  if (is.null(input_data)) {
    return(invisible(NULL))
  }
  
  # Get output directory
  output_dir <- get_save_path()
  if (is.null(output_dir)) {
    cat("Operation cancelled.\n")
    return(invisible(NULL))
  }
  
  # Process each row in the input data
  cat("\nProcessing", nrow(input_data), "projects from the input file...\n")
  
  for (i in 1:nrow(input_data)) {
    tryCatch({
      # Extract data from the row
      row <- input_data[i, ]
      project_name <- row$Project
      future_capex <- as.numeric(row$Future_Capex)
      lom_ounces <- as.numeric(row$LOM_Ounces)
      ounces_mined <- as.numeric(row$Ounces_Mined)
      
      cat("\nProcessing Project:", project_name, "\n")
      cat("  Future Capex: $", format(future_capex, big.mark = ",", scientific = FALSE), "\n", sep = "")
      cat("  LOM Ounces: ", format(lom_ounces, big.mark = ",", scientific = FALSE), "\n", sep = "")
      cat("  Ounces Mined: ", format(ounces_mined, big.mark = ",", scientific = FALSE), "\n", sep = "")
      
      # Create project-specific output directory
      project_dir <- file.path(output_dir, project_name)
      dir.create(project_dir, recursive = TRUE, showWarnings = FALSE)
      
      # Perform sensitivity analysis
      results <- sensitivity_analysis(future_capex, lom_ounces, ounces_mined)
      amort_results <- results$amort_results
      expense_results <- results$expense_results
      
      # Save results to Excel
      excel_filename <- paste0(project_name, "_sensitivity_analysis.xlsx")
      excel_path <- file.path(project_dir, excel_filename)
      
      # Create a list of data frames for the Excel file
      sheets <- list(
        "Amortization Rates" = amort_results,
        "Expected Expenses" = expense_results,
        "Input Summary" = data.frame(
          Parameter = c("Project", "Future Capex", "LOM Ounces", "Ounces Mined"),
          Value = c(project_name, future_capex, lom_ounces, ounces_mined)
        )
      )
      
      # Write to Excel
      writexl::write_xlsx(sheets, excel_path)
      cat("  Excel results saved to:", excel_path, "\n")
      
      # Create and save amortization rate heatmap
      create_heatmap(
        amort_results,
        paste0("Amortization Rate Sensitivity Analysis: ", project_name, " ($/ounce)"),
        paste0(project_name, "_amortization_sensitivity.png"),
        project_dir
      )
      
      # Create and save expected expense heatmap
      create_heatmap(
        expense_results,
        paste0("Expected Expense Sensitivity: ", project_name, " (", 
               format(ounces_mined, big.mark = ",", scientific = FALSE), " Ounces Mined)"),
        paste0(project_name, "_expense_sensitivity.png"),
        project_dir
      )
      
      cat("  Sensitivity analysis completed for", project_name, "\n")
      
    }, error = function(e) {
      cat("Error processing row", i, "(Project:", input_data$Project[i], "):", e$message, "\n")
    })
  }
  
  cat("\nBatch processing completed!\n")
}

# Run the main function
main()

File: rerun.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using OfficeOpenXml;
using ScottPlot;
using ScottPlot.Palettes;
using System.Drawing;

// For EPPlus licensing
[assembly: OfficeOpenXml.ExcelPackage.LicenseContext(OfficeOpenXml.LicenseContext.NonCommercial)]

namespace SensitivityAnalysis
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Batch Sensitivity Analysis for Amortization Rate and Expected Expenses");
            Console.WriteLine("--------------------------------------------------------------------");

            // Get input Excel file path
            string inputFile = GetFilePath("Enter the path to your input Excel file: ");
            if (string.IsNullOrEmpty(inputFile))
            {
                Console.WriteLine("Operation cancelled.");
                return;
            }

            // Read input data
            List<ProjectData> projects = ReadInputExcel(inputFile);
            if (projects == null || projects.Count == 0)
            {
                return;
            }

            // Get output directory
            string outputDir = GetSavePath();
            if (string.IsNullOrEmpty(outputDir))
            {
                Console.WriteLine("Operation cancelled.");
                return;
            }

            // Process each project
            Console.WriteLine($"\nProcessing {projects.Count} projects from the input file...");

            foreach (var project in projects)
            {
                try
                {
                    Console.WriteLine($"\nProcessing Project: {project.ProjectName}");
                    Console.WriteLine($"  Future Capex: ${project.FutureCapex:N2}");
                    Console.WriteLine($"  LOM Ounces: {project.LomOunces:N2}");
                    Console.WriteLine($"  Ounces Mined: {project.OuncesMined:N2}");

                    // Create project-specific output directory
                    string projectDir = Path.Combine(outputDir, project.ProjectName);
                    Directory.CreateDirectory(projectDir);

                    // Perform sensitivity analysis
                    SensitivityResults results = RunSensitivityAnalysis(
                        project.FutureCapex, project.LomOunces, project.OuncesMined);

                    // Save results to Excel
                    string excelPath = Path.Combine(projectDir, $"{project.ProjectName}_sensitivity_analysis.xlsx");
                    SaveToExcel(results, project, excelPath);
                    Console.WriteLine($"  Excel results saved to: {excelPath}");

                    // Create and save amortization rate heatmap
                    string amortHeatmapPath = Path.Combine(projectDir, $"{project.ProjectName}_amortization_sensitivity.png");
                    CreateHeatmap(
                        results.AmortizationRates,
                        results.PercentageLabels,
                        $"Amortization Rate Sensitivity Analysis: {project.ProjectName} ($/ounce)",
                        amortHeatmapPath);
                    Console.WriteLine($"  Amortization heatmap saved to: {amortHeatmapPath}");

                    // Create and save expected expense heatmap
                    string expenseHeatmapPath = Path.Combine(projectDir, $"{project.ProjectName}_expense_sensitivity.png");
                    CreateHeatmap(
                        results.ExpectedExpenses,
                        results.PercentageLabels,
                        $"Expected Expense Sensitivity: {project.ProjectName} ({project.OuncesMined:N0} Ounces Mined)",
                        expenseHeatmapPath);
                    Console.WriteLine($"  Expense heatmap saved to: {expenseHeatmapPath}");

                    Console.WriteLine($"  Sensitivity analysis completed for {project.ProjectName}");
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error processing project {project.ProjectName}: {ex.Message}");
                }
            }

            Console.WriteLine("\nBatch processing completed!");
        }

        static string GetFilePath(string prompt)
        {
            while (true)
            {
                Console.Write(prompt);
                string filePath = Console.ReadLine().Trim();

                if (string.IsNullOrEmpty(filePath))
                {
                    return null;
                }

                if (File.Exists(filePath))
                {
                    return filePath;
                }
                else
                {
                    Console.WriteLine($"Error: File '{filePath}' does not exist.");
                }
            }
        }

        static string GetSavePath()
        {
            while (true)
            {
                Console.Write("\nEnter the folder path to save results (or press Enter to use current directory): ");
                string saveDir = Console.ReadLine().Trim();

                // Use current directory if input is empty
                if (string.IsNullOrEmpty(saveDir))
                {
                    return Directory.GetCurrentDirectory();
                }

                // Check if the directory exists
                if (Directory.Exists(saveDir))
                {
                    return saveDir;
                }
                else
                {
                    Console.Write($"Directory '{saveDir}' doesn't exist. Create it? (y/n): ");
                    string createDir = Console.ReadLine().ToLower();
                    if (createDir == "y")
                    {
                        try
                        {
                            Directory.CreateDirectory(saveDir);
                            Console.WriteLine($"Created directory: {saveDir}");
                            return saveDir;
                        }
                        catch (Exception ex)
                        {
                            Console.WriteLine($"Error creating directory: {ex.Message}");
                        }
                    }
                    else
                    {
                        Console.WriteLine("Please enter a valid directory path.");
                    }
                }
            }
        }

        static List<ProjectData> ReadInputExcel(string filePath)
        {
            try
            {
                ExcelPackage.LicenseContext = LicenseContext.NonCommercial;

                using (var package = new ExcelPackage(new FileInfo(filePath)))
                {
                    var worksheet = package.Workbook.Worksheets[0]; // First worksheet
                    int rowCount = worksheet.Dimension.Rows;
                    int colCount = worksheet.Dimension.Columns;

                    // Validate headers
                    string[] requiredColumns = { "Project", "Future_Capex", "LOM_Ounces", "Ounces_Mined" };
                    Dictionary<string, int> columnIndices = new Dictionary<string, int>();

                    for (int col = 1; col <= colCount; col++)
                    {
                        string headerValue = worksheet.Cells[1, col].Text;
                        if (requiredColumns.Contains(headerValue))
                        {
                            columnIndices[headerValue] = col;
                        }
                    }

                    // Check if all required columns exist
                    var missingColumns = requiredColumns.Where(col => !columnIndices.ContainsKey(col)).ToList();
                    if (missingColumns.Any())
                    {
                        Console.WriteLine($"Error: Missing required columns in the Excel file: {string.Join(", ", missingColumns)}");
                        Console.WriteLine("Please ensure your Excel file has columns named: Project, Future_Capex, LOM_Ounces, Ounces_Mined");
                        return null;
                    }

                    // Read data rows
                    var projects = new List<ProjectData>();
                    for (int row = 2; row <= rowCount; row++) // Start from row 2 (skip header)
                    {
                        string projectName = worksheet.Cells[row, columnIndices["Project"]].Text;
                        
                        // Try to parse numeric values
                        double futureCapex, lomOunces, ouncesMined;
                        
                        if (!double.TryParse(worksheet.Cells[row, columnIndices["Future_Capex"]].Text, out futureCapex) ||
                            !double.TryParse(worksheet.Cells[row, columnIndices["LOM_Ounces"]].Text, out lomOunces) ||
                            !double.TryParse(worksheet.Cells[row, columnIndices["Ounces_Mined"]].Text, out ouncesMined))
                        {
                            Console.WriteLine($"Warning: Row {row} contains invalid numeric data. Skipping this project.");
                            continue;
                        }

                        if (string.IsNullOrEmpty(projectName))
                        {
                            Console.WriteLine($"Warning: Row {row} is missing a project name. Skipping this project.");
                            continue;
                        }

                        projects.Add(new ProjectData
                        {
                            ProjectName = projectName,
                            FutureCapex = futureCapex,
                            LomOunces = lomOunces,
                            OuncesMined = ouncesMined
                        });
                    }

                    if (projects.Count == 0)
                    {
                        Console.WriteLine("Error: No valid data rows found after cleaning.");
                        return null;
                    }

                    return projects;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error reading Excel file: {ex.Message}");
                return null;
            }
        }

        static SensitivityResults RunSensitivityAnalysis(double baseFutureCapex, double baseLomOunces, 
                                                        double ouncesMined, double variation = 0.20, int steps = 5)
        {
            // Calculate variation percentages
            int numSteps = (int)(2 * variation / (variation / steps)) + 1;
            double[] percentages = new double[numSteps];
            for (int i = 0; i < numSteps; i++)
            {
                percentages[i] = -variation + i * (variation / steps * 2);
            }

            // Create arrays for future capex and LOM ounces variations
            double[] futureCapexVariations = new double[percentages.Length];
            double[] lomOuncesVariations = new double[percentages.Length];
            for (int i = 0; i < percentages.Length; i++)
            {
                futureCapexVariations[i] = baseFutureCapex * (1 + percentages[i]);
                lomOuncesVariations[i] = baseLomOunces * (1 + percentages[i]);
            }

            // Create percentage labels for the table
            string[] percentageLabels = new string[percentages.Length];
            for (int i = 0; i < percentages.Length; i++)
            {
                percentageLabels[i] = $"{(int)(percentages[i] * 100)}%";
            }

            // Initialize matrices for the results
            double[,] amortMatrix = new double[percentages.Length, percentages.Length];
            double[,] expenseMatrix = new double[percentages.Length, percentages.Length];

            // Fill the matrices with amortization rates and expected expenses
            for (int i = 0; i < percentages.Length; i++)
            {
                for (int j = 0; j < percentages.Length; j++)
                {
                    // Calculate amortization rate
                    if (lomOuncesVariations[j] == 0)
                    {
                        amortMatrix[i, j] = double.NaN; // Avoid division by zero
                    }
                    else
                    {
                        amortMatrix[i, j] = futureCapexVariations[i] / lomOuncesVariations[j];
                    }

                    // Calculate expected expenses
                    expenseMatrix[i, j] = amortMatrix[i, j] * ouncesMined;
                }
            }

            return new SensitivityResults
            {
                AmortizationRates = amortMatrix,
                ExpectedExpenses = expenseMatrix,
                PercentageLabels = percentageLabels
            };
        }

        static void SaveToExcel(SensitivityResults results, ProjectData project, string filePath)
        {
            ExcelPackage.LicenseContext = LicenseContext.NonCommercial;

            using (var package = new ExcelPackage())
            {
                // Add Amortization Rates worksheet
                var amortSheet = package.Workbook.Worksheets.Add("Amortization Rates");
                
                // Add row and column headers
                for (int i = 0; i < results.PercentageLabels.Length; i++)
                {
                    amortSheet.Cells[1, i + 2].Value = results.PercentageLabels[i];
                    amortSheet.Cells[i + 2, 1].Value = results.PercentageLabels[i];
                }
                
                // Add header titles
                amortSheet.Cells[1, 1].Value = "Future Capex / LOM Ounces";
                
                // Add data
                for (int i = 0; i < results.PercentageLabels.Length; i++)
                {
                    for (int j = 0; j < results.PercentageLabels.Length; j++)
                    {
                        amortSheet.Cells[i + 2, j + 2].Value = results.AmortizationRates[i, j];
                        amortSheet.Cells[i + 2, j + 2].Style.Numberformat.Format = "0.00";
                    }
                }
                
                // Format headers
                using (var range = amortSheet.Cells[1, 1, 1, results.PercentageLabels.Length + 1])
                {
                    range.Style.Font.Bold = true;
                    range.Style.Fill.PatternType = OfficeOpenXml.Style.ExcelFillStyle.Solid;
                    range.Style.Fill.BackgroundColor.SetColor(Color.LightGray);
                }
                
                using (var range = amortSheet.Cells[1, 1, results.PercentageLabels.Length + 1, 1])
                {
                    range.Style.Font.Bold = true;
                    range.Style.Fill.PatternType = OfficeOpenXml.Style.ExcelFillStyle.Solid;
                    range.Style.Fill.BackgroundColor.SetColor(Color.LightGray);
                }
                
                amortSheet.Cells.AutoFitColumns();

                // Add Expected Expenses worksheet
                var expenseSheet = package.Workbook.Worksheets.Add("Expected Expenses");
                
                // Add row and column headers
                for (int i = 0; i < results.PercentageLabels.Length; i++)
                {
                    expenseSheet.Cells[1, i + 2].Value = results.PercentageLabels[i];
                    expenseSheet.Cells[i + 2, 1].Value = results.PercentageLabels[i];
                }
                
                // Add header titles
                expenseSheet.Cells[1, 1].Value = "Future Capex / LOM Ounces";
                
                // Add data
                for (int i = 0; i < results.PercentageLabels.Length; i++)
                {
                    for (int j = 0; j < results.PercentageLabels.Length; j++)
                    {
                        expenseSheet.Cells[i + 2, j + 2].Value = results.ExpectedExpenses[i, j];
                        expenseSheet.Cells[i + 2, j + 2].Style.Numberformat.Format = "0.00";
                    }
                }
                
                // Format headers
                using (var range = expenseSheet.Cells[1, 1, 1, results.PercentageLabels.Length + 1])
                {
                    range.Style.Font.Bold = true;
                    range.Style.Fill.PatternType = OfficeOpenXml.Style.ExcelFillStyle.Solid;
                    range.Style.Fill.BackgroundColor.SetColor(Color.LightGray);
                }
                
                using (var range = expenseSheet.Cells[1, 1, results.PercentageLabels.Length + 1, 1])
                {
                    range.Style.Font.Bold = true;
                    range.Style.Fill.PatternType = OfficeOpenXml.Style.ExcelFillStyle.Solid;
                    range.Style.Fill.BackgroundColor.SetColor(Color.LightGray);
                }
                
                expenseSheet.Cells.AutoFitColumns();

                // Add Input Summary worksheet
                var summarySheet = package.Workbook.Worksheets.Add("Input Summary");
                
                summarySheet.Cells[1, 1].Value = "Parameter";
                summarySheet.Cells[1, 2].Value = "Value";
                
                summarySheet.Cells[2, 1].Value = "Project";
                summarySheet.Cells[2, 2].Value = project.ProjectName;
                
                summarySheet.Cells[3, 1].Value = "Future Capex";
                summarySheet.Cells[3, 2].Value = project.FutureCapex;
                summarySheet.Cells[3, 2].Style.Numberformat.Format = "#,##0.00";
                
                summarySheet.Cells[4, 1].Value = "LOM Ounces";
                summarySheet.Cells[4, 2].Value = project.LomOunces;
                summarySheet.Cells[4, 2].Style.Numberformat.Format = "#,##0.00";
                
                summarySheet.Cells[5, 1].Value = "Ounces Mined";
                summarySheet.Cells[5, 2].Value = project.OuncesMined;
                summarySheet.Cells[5, 2].Style.Numberformat.Format = "#,##0.00";
                
                // Format headers
                using (var range = summarySheet.Cells[1, 1, 1, 2])
                {
                    range.Style.Font.Bold = true;
                    range.Style.Fill.PatternType = OfficeOpenXml.Style.ExcelFillStyle.Solid;
                    range.Style.Fill.BackgroundColor.SetColor(Color.LightGray);
                }
                
                summarySheet.Cells.AutoFitColumns();

                // Save the Excel file
                package.SaveAs(new FileInfo(filePath));
            }
        }

        static void CreateHeatmap(double[,] data, string[] labels, string title, string filePath)
        {
            var plt = new ScottPlot.Plot(800, 600);

            // Create heatmap
            var heatmap = plt.AddHeatmap(data, lockScales: false);
            heatmap.ShowAxisLabels = true;
            heatmap.XTickLabels = labels;
            heatmap.YTickLabels = labels;
            
            // Add a colorbar
            plt.AddColorbar(heatmap);

            // Set plot properties
            plt.Title(title);
            plt.XLabel("LOM Ounces Variation");
            plt.YLabel("Future Capex Variation");
            
            // Customize axis ticks
            plt.XAxis.TickLabelStyle(rotation: 45);
            
            // Display values in cells
            for (int i = 0; i < data.GetLength(0); i++)
            {
                for (int j = 0; j < data.GetLength(1); j++)
                {
                    // Invert the i-coordinate for plotting due to how heatmaps are displayed
                    string text = $"{data[i, j]:F2}";
                    plt.AddText(text, j, i, size: 12, 
                                color: data[i, j] > (data.Cast<double>().Average()) ? Color.White : Color.Black);
                }
            }

            // Save to file
            plt.SaveFig(filePath);
        }
    }

    class ProjectData
    {
        public string ProjectName { get; set; }
        public double FutureCapex { get; set; }
        public double LomOunces { get; set; }
        public double OuncesMined { get; set; }
    }

    class SensitivityResults
    {
        public double[,] AmortizationRates { get; set; }
        public double[,] ExpectedExpenses { get; set; }
        public string[] PercentageLabels { get; set; }
    }
}
